packages/test-utils/src/reporter.ts

```diff
@@ -920,52 +920,108 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
 
       let uploadableResults = this._uploadableResults.get(result.testId);
       if (!uploadableResults) {
-        uploadableResults = [];
+        uploadableResults = {
+          executions: {},
+          aggregateStatus: undefined,
+          didUploadStatuses: {
+            passed: false,
+            failed: false,
+          },
+        };
         this._uploadableResults.set(result.testId, uploadableResults);
       }
-      uploadableResults.push(result);
+      let executions = uploadableResults.executions[result.executionGroupId];
+      if (!executions) {
+        executions = [];
+        uploadableResults.executions[result.executionGroupId] = executions;
+      }
+      executions.push(result);
 
       if (result.result === "passed" || result.attempt >= result.maxAttempts) {
-        this._enqueueUploads(uploadableResults);
+        this._enqueueUploads(uploadableResults, result.executionGroupId);
       }
     }
   }
 
-  private _enqueueUploads(results: UploadableTestRunResult<TRecordingMetadata>[]) {
-    const latestResult = results[results.length - 1];
-    assert(!!latestResult, "Expected at least one result in the list");
+  private _enqueueUploads(
+    result: UploadableTestResult<TRecordingMetadata>,
+    executionGroupId: string
+  ) {
+    const executions = result.executions[executionGroupId];
+    const latestExecution = last(executions);
+    assert(!!latestExecution, "Expected at least one execution in the list");
+
+    let toUpload: typeof executions = [];
 
-    let toUpload: typeof results | undefined;
+    const aggregateStatus = this._assignAggregateStatus(result, executions);
 
-    switch (this._uploadStatusThreshold) {
-      case "all":
-        toUpload = this._minimizeUploads ? [latestResult] : results;
+    switch (aggregateStatus) {
+      case "failed": {
+        if (!this._minimizeUploads) {
+          result.didUploadStatuses.failed = true;
+          toUpload.push(...executions);
+          break;
+        }
+        if (result.didUploadStatuses.failed) {
+          break;
+        }
+        result.didUploadStatuses.failed = true;
+        toUpload.push(latestExecution);
```

Comment:
I think we need to check if status threshold is "none" here before uploading?

```suggestion
        if (this._uploadStatusThreshold !== "none") {
          if (this._minimizeUploads) {
            if (!result.didUploadStatuses.failed) {
              result.didUploadStatuses.failed = true;
              toUpload.push(latestExecution);
            }
          } else {
            result.didUploadStatuses.failed = true;
            toUpload.push(...executions);
          }
        }
```

TBH I guess we could just return early from this function if that's the status because the "flaky" branch below also needs to respect that condition.